[IMP] runbot_merge: recombine remaining batches after identifying a culprit PR

MiquelRForgeFlow opened the pull request:
**Background (current behavior):**
When a staging with multiple batches fails, runbot_merge splits the set in two halves and stages them in sequence. Only the failing half is split again (bisected); a passing half is staged and merged as a unit. Eventually, if bisecting reaches a single-batch staging that fails, the culprit PR is identified and marked as failed. After that point, the remaining PRs from the original split group continue through the existing split queue (halves and further splits only if those fail). There is no step to recombine all unaffected PRs into a single staging once the culprit is known.

**Problem:**
Even when there is only one bad PR, the remaining good PRs may still be processed across several stagings (depending on how the queued splits unfold), leading to extra CI cycles and slower throughput.

**Proposed change:**
As soon as a culprit PR is identified in a single-batch staging:

1. Mark that PR as failed (existing behavior).
~~2. Cancel any pending stagings from the same split source and remove sibling splits for that source.~~
3. Create a new split containing all remaining active batches from the original group (excluding the culprit and closed/merged ones).
4. Let the scheduler stage this recombined batch together.

**Why this helps:**
- Fewer CI runs: In the common “one culprit” scenario, we avoid cascading stagings of separate halves.
- Faster merges: Good PRs get merged together promptly.
- Safe: If another culprit or a problematic interaction exists, the recombined staging will fail and the standard bisect logic will isolate it.

**Implementation notes:**

- New helper `recombine_remaining_after_culprit(pr)` on `runbot_merge.stagings`. Finds sibling splits by `source_id`, filters remaining active batches, cancels pending sibling stagings, unlinks sibling splits for the same source, then creates a single recombined split with the remaining batches.
- Set `_order = 'id desc'` on `runbot_merge.split` so the newly created recombined split is picked first by `try_staging()`.

**How to test (manual):**
1. Prepare 4 batches `B1..B4` on the same target. Make `B1` fail in isolation; others pass.
2. Trigger a staging with all 4 → it fails → it is split into `[B1,B2]` and `[B3,B4]`.
3. Process the left split; bisect isolates `B1` as culprit (single-batch failure).
4. Verify logs show: “recombine remaining N batches …” and that a new split is created with `B2,B3,B4`.
5. Next scheduler run should stage `B2,B3,B4` together; CI passes → they merge together.
6. Repeat with a “two culprits” scenario to confirm the process repeats until all bad PRs are isolated.

**Backward compatibility & failure modes:**
- Worst case (multiple culprits/interactions): behavior naturally falls back to standard bisect.
- No change to approvals/labels; only scheduling is optimized.
- Recombined staging creation is idempotent per split source cycle.

If you’d like, I can add a short “Changelog entry”.

MiquelRForgeFlow committed:
[IMP] runbot_merge: recombine remaining batches after culprit PR

When a multi-batch staging fails and a culprit PR is later identified
in a single-batch staging, cancel pending sibling stagings/splits from
the same split root and create a new split with all remaining active
batches. This lets the scheduler stage them together again, reducing CI
cycles and accelerating merges while preserving safety (bisect still
applies if more culprits/interactions remain).

MiquelRForgeFlow commented:
@Xavier-Do @d-fence Hope you like it.

Xavier-Do commented:
Hello @MiquelRForgeFlow 

Thanks for contributing

This is more a topic for @xmo-odoo but it looks like it wouldn't work well with random errors.

MiquelRForgeFlow commented:
@Xavier-Do to be more safe, I removed the cancelling of any pending staging from the same split source and sibling splits for that source. The new combined split should be executed regardless because the `_order` set on `runbot_merge.split`.

BTW, I asked ChatGPT how to handle better the random errors (flakes) and it proposes two ways:
1. Culprit confirmation with a short retry (“canary retry”): Before marking a PR as failed (when you detect it as single-batch), make a unique retry of that single-batch.
2. Unique retry of the recombined / single-batched when there is no clear culprit.

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2304,6 +2304,41 @@ def fail(self, message, prs=None):
+        culprit_batch = culprit_pr.batch_id
+        if not culprit_batch:
+            _logger.info("No culprit batch found for PR %s; skipping recombine", culprit_pr)
+            return False
```
This does not make sense, the staging only accesses the PRs through batches.

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2304,6 +2304,41 @@ def fail(self, message, prs=None):
+            _logger.info("No sibling splits for root staging %s; skipping recombine", source)
+            return False
+
+        candidate_batches = sibling_splits.mapped('batch_ids').filtered(lambda b: b.id != culprit_batch.id and b.active)
```
Splits are transient and deleted when staged
https://github.com/odoo/runbot/blob/cd13f03bda397cb6b3cdd6e3c3e222513c423315/runbot_merge/models/stagings_create.py#L93-L97
so a staged batch can not be part of a split. And if extant splits are regrouped into a new split, the old splits need to be deleted

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2304,6 +2304,41 @@ def fail(self, message, prs=None):
+        })
+        _logger.info(
+            "Recombined %d batches after culprit PR#%s into %s (source=%s)",
+            len(candidate_batches), culprit_pr.number, new_split, source
```
A PR number is wholly insufficient to identify the PR, as they're only unique within a repository. The mergebot overrides `PullRequests.display_name` to write out the fully qualified short form of the PR (`{owner}/{name}#{number}`)

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2304,6 +2304,41 @@ def fail(self, message, prs=None):
+        sibling_splits = Split.search([('source_id', '=', source.id)])
+        if not sibling_splits:
+            _logger.info("No sibling splits for root staging %s; skipping recombine", source)
+            return False
```
Almost certainly unnecessarily complicated, Split.parent_id only exists to create a link between stagings, which itself only exists with the idea that we might one day use it to try and better identify staging false positives (and auto-scan them or something).

At a higher level of abstraction, if there are splits for a target they all come from the same original staging.

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2304,6 +2304,41 @@ def fail(self, message, prs=None):
+            "Recombined %d batches after culprit PR#%s into %s (source=%s)",
+            len(candidate_batches), culprit_pr.number, new_split, source
+        )
+        return True
```
Although the return value of `try_splitting` is not used anyway and should probably be removed, this is incorrect. The semantics of `try_splitting` is that it returns `True` if it managed to split the staging. Since the recombination of extant splits can only come at the consequence of a single-batch staging, the staging can not be split.

xmo-odoo commented on runbot_merge/models/pull_requests.py:
```diff
@@ -2357,11 +2392,15 @@ def try_splitting(self):
                 self.fail("%s%s" % (reason, viewmore), pr)
+                try:
+                    return self.recombine_remaining_after_culprit(pr)
+                except Exception:
```
Why?

xmo-odoo reviewed (commented):
I added both my understanding of the original version (a depth-first search of a culprit PR) and this version (keep the breadth-first search but recombine as soon as a "culprit" is found) to a small simulator I worked on during the odoo days. I found the original to be way better than I was assuming, while the simple recombination is not really an improvemend over the baseline version:

title              |  total | merged | failed | nondeterministic
-------------------|--------:|--------:|--------:|------------------:
batched            | 236000 | 233380 |   2222 |              398
batched_rejoin     | 238608 | 236005 |   2210 |              393
batched_depthfirst | 254902 | 252245 |   2377 |              280

(this is with a bunch of assumptions as the simulator is pretty simplistic, and over 5 years of simulation-time).

However,

- the branch for the mergebot is 17.0
- the mergebot has a fair number of tests, and as both this and the previous significantly impact the scheduling of splits I doubt the various tests which involve such splits still pass with this

MiquelRForgeFlow commented:
@xmo-odoo comments attended.

MiquelRForgeFlow commented:
@xmo-odoo is there something else to do?

xmo-odoo commented:
Finally got around to merging this, kinda, as well as the original form (doing depth-first search of failing PRs).  Thanks.

I removed much of the logging as I think this version is excessive, and put the recombination in `fail`: it doesn't matter that we can't really identify the specific PR in a multi-PR batch, for this it matters that we have failed a single batch and assume that batch is the deterministic source of errors. Same for timeout errors, if a batch deterministically causes timeouts then the rest is absolved.

I also put in a tunable so that the splits can just be deleted instead of recombined, if false positive rates are low then this should mostly yield higher throughput, and staging priority instability should mostly be fixed since 6fa190b08f63123cadf30afd84ab6651d30032fe so keeping splits around for that purpose (not having a batch overtaken by a latter one on restage) should mostly be unnecessary.
